Fix: WooCommerce Blocks causing malfunction of the navigation block on WordPress 6.3

nefeline opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

This PR solves the conflict preventing the core's navigation block from being correctly previewed in the editor while having WooCommerce 7.9.0 running alongside WordPress version 6.3.

<!-- Reference any related issues or PRs here -->

Fixes #10259

## Investigation and Details

This issue was introduced on WooCommerce Blocks version `10.3.0` by https://github.com/woocommerce/woocommerce-blocks/pull/9468 . That PR was created to solve an issue where the Products (Beta) block would not be available in the editor inserter when running `npm start`.

After reverting the changes that caused the conflict, and ensuring the mini-cart could still be inserted in the editor (as that block was also relying on the same `isSiteEditorPage` hook), @Aljullu helped me out with circumventing other limitations found, such as ensuring we can still hide the Inherit query control in the post/page editor and insert the Products (Beta) block while running `npm start`.

@Aljullu also brought up this work by @nerrad that fixed some issues related to `useSelect()` in WC core which could be related to the problem we are facing here and served as an inspiration: https://github.com/woocommerce/woocommerce/pull/37641

If anyone has a detailed explanation for the cause of this issue I would highly appreciate it: is it because we were calling `useSelect()` directly? If that's the case, why? Is it something else entirely?

## Important Note

In its current shape and form, this PR fixes the problem for all users while still allowing us to insert the impacted blocks on `npm start`, the only caveat is that the navigation block itself still throws the same error _exclusively_ on `npm start`. 

Since this limitation impacts developers solely, and we need this fix to be included in the core of Woo (on version 8.0.0) ASAP as WordPress 6.3 will be released on August 8th, a follow-up issue is being opened to address this limitation separately.

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

### Screenshots

<!-- If your change has a visual component, add a screenshot here. A "before" screenshot would also be helpful. -->

| Before | After |
| ------ | ----- |
| <img width="913" alt="Screenshot 2023-07-23 at 11 15 40" src="https://github.com/woocommerce/woocommerce-blocks/assets/15730971/d473a148-bfa3-4622-b714-bde0e689df35"> | <img width="912" alt="Screenshot 2023-07-23 at 11 14 09" src="https://github.com/woocommerce/woocommerce-blocks/assets/15730971/4c771973-32c1-4835-851b-09aa4bb40653"> |

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Ensure you have a local install up and running with the release candidate version of WordPress 6.3. You can do so by downloading it via WP-CLI: `wp core update --version=6.3-RC1 --force` followed by `wp core update-db`. Alternatively, you can also download the zip for the release [over here](https://wordpress.org/wordpress-6.3-RC1.zip).
2. Create a new post
3. Insert the navigation block (if you are testing this locally, make sure assets are compiled via `npm run build`)
4. Make sure you can correctly preview it without any problems, as demonstrated in the screenshot in the description of this PR.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

### Changelog

> Fix: Resolve the conflict preventing the navigation block from being correctly previewed in the editor while having WooCommerce enabled.

nefeline committed:
Revert cabbaa270

nefeline committed:
Ensure the mini-cart works as expected when running on npm start

Aljullu committed:
Fix Products (Beta) block not appearing in the post/page editor

Aljullu committed:
Fix Inherit Query control appearing in the post editor on npm run start

Aljullu committed:
Fix Products (Beta) block not appearing in the post/page editor (II)

nefeline committed:
Merge trunk and resolve conflicts.

nerrad commented on package.json:
```diff
@@ -27,8 +27,7 @@
 		"./assets/js/blocks/product-query/variations/**.tsx",
 		"./assets/js/blocks/product-query/index.tsx",
-		"./assets/js/blocks/product-query/inspector-controls.tsx",
-		"./assets/js/templates/revert-button/index.tsx"
```
What was this path removed from the side-effects configuration?

nerrad replied:
Since `revert-button/index.tsx` is the only file that has a `@wordpress/edit-site` dependency. I suspect this is actually the cause of the issue, especially since it is [imported as a root asset](https://github.com/woocommerce/woocommerce-blocks/blob/2cddfcdc3001afd412450809fbc48a11046b90f9/assets/js/index.js#L8)! It should only be imported/utilized in the site editor environment.

nefeline replied:
> Since revert-button/index.tsx is the only file that has a @wordpress/edit-site dependency. I suspect this is actually the cause of the issue, especially since it is [imported as a root asset](https://github.com/woocommerce/woocommerce-blocks/blob/2cddfcdc3001afd412450809fbc48a11046b90f9/assets/js/index.js#L8)! It should only be imported/utilized in the site editor environment.

Oh, nice catch: that sounds like the cause of the problem here indeed!

> What was this path removed from the side-effects configuration?

The revert button works as expected with and without having `"./assets/js/templates/revert-button/index.tsx"` in this side-effect: it looks like this was added so it can run on production builds. Tagging @gigitux for confirmation.

gigitux replied:
Thanks for working on this!

> Since revert-button/index.tsx is the only file that has a @wordpress/edit-site dependency. I suspect this is actually the cause of the issue, especially since it is [imported as a root asset](https://github.com/woocommerce/woocommerce-blocks/blob/2cddfcdc3001afd412450809fbc48a11046b90f9/assets/js/index.js#L8)! It should only be imported/utilized in the site editor environment.

I was aware of this issue, I tried to use a dynamic import, but it didn't work. [I wrote a message on the Gutenberg PR](https://github.com/WordPress/gutenberg/pull/50257#issuecomment-1549279797). I have no idea how to fix this in a good way, but I'm happy to help if needed!

> The revert button works as expected with and without having "./assets/js/templates/revert-button/index.tsx" in this side-effect: it looks like this was added so it can run on production builds. Tagging @gigitux for confirmation.

True!

nerrad replied:
> The revert button works as expected with and without having "./assets/js/templates/revert-button/index.tsx" in this side-effect: it looks like this was added so it can run on production builds. Tagging @gigitux for confirmation.

That's odd, the described behaviour in this PR of the navigation block issue being resolved in production builds but _not_ in development builds is consistent with the import being stripped out of production builds because it is considered a side-effect. This should be easily verifiable by checking to see if `wp.editSite` is available in the post editor with a production build of Woo Blocks.

nefeline replied:
You are right; I may have hit caching on my initial tests. I just had an idea to circumvent this: I will try to conditionally load the necessary script via PHP rather than JS, as I suspect it's not possible to truly conditionally import a module directly in the JavaScript execution flow.

nerrad commented on assets/js/blocks/mini-cart/edit.tsx:
```diff
@@ -80,7 +82,7 @@ const Edit = ( { attributes, setAttributes }: Props ): ReactElement => {
 	} );
 
-	const isSiteEditor = isSiteEditorPage( select( 'core/edit-site' ) );
+	const isSiteEditor = postType === undefined || postId === undefined;
```
Why are you just using the presence of a `postType` or `postId` value as an indicator of whether this is in the site editor environment?

nefeline replied:
There are a few places on GB where they use the same, such as [within the post-comments-form](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/post-comments-form/form.js#L63), so used that as an inspiration, but I agree this doesn't seem to be a rock-solid foundation to build a house on. Is there a recommended approach for checking if a site is in editor environment?

Aljullu replied:
Yup, as @nefeline says, we used the Post Comments Form block as a reference here. And even though the variable is named `isSiteEditor`, I think what's relevant here is whether a template is being edited or not.

I guess we could rewrite this code to use the same approach that we are introducing for Product Query:

```TS
const isTemplate = useSelect( ( select ) => {
	const store = select( 'core/edit-site' );
	const editedPostType = store?.getEditedPostType();

	return editedPostType === 'wp_template';
}, [] );
```

Does that sound like a good approach?

Aljullu replied:
I went ahead and made the change in e9defb0358f9b0807cbc32225efaa949adebdf9e.

nefeline replied:
> Does that sound like a good approach?

👍

nerrad commented on assets/js/blocks/product-query/variations/product-query.tsx:
```diff
@@ -65,16 +64,24 @@ const registerProductsBlock = ( attributes: QueryBlockAttributes ) => {
+	if ( isTemplate ) {
+		let currentTemplateId: string | undefined;
+
+		subscribe( () => {
+			const previousTemplateId = currentTemplateId;
+
+			currentTemplateId = store?.getEditedPostId();
+
+			if (
+				previousTemplateId === currentTemplateId ||
+				typeof currentTemplateId === 'undefined'
+			) {
+				return;
+			}
```
The current logic has a heavy assumption on `isTemplate` either being fully initialized when `registerProductsBlock` is called or never changing in the app's lifecycle. That's not a good assumption to make here. Why not have this check within the subscribe callback?

Aljullu replied:
Good point. I refactored this part of the code in 2057f0f920e31dd6e4f1ffebf7f322c798b28bae, I think it should be more resilient now and (slightly) more performant.

nerrad commented on assets/js/templates/revert-button/index.tsx:
```diff
@@ -140,12 +140,7 @@ let currentTemplateId: string | undefined;
-	}
-
-	currentTemplateId = store?.getEditedPostId();
+	currentTemplateId = store.getEditedPostId();
```
It's safe to assume that this component will always be present in a site-editor environment?

nefeline replied:
You are right, it's better to not make this assumption, updated 👍 !

nerrad reviewed (changes requested):
One clue I found related to the original issue is that with WooCommerce not active, `wp.editSite` does not have a value in the post editor whereas when it is active it does. This indicates that something in WooCommerce is loaded in the post editor environment with a dependency on `@wordpress/site-editor` and this ultimately causes the error.

I notice that in your changes, you removed the `templates/revert-button/index.tsx` path from the side-effects array configuration in `package.json`. That file imports the `@wordpress/site-editor` package (and is the only file in all of Woo to do so). Likely not a coincidence and *also* indicates that the revert button functionality might not be working as expected in production builds (I don't see it in your testing notes).

nefeline committed:
Don't make the assumption that getEditedPostId will always be available in the site-editor.

nefeline commented:
> I notice that in your changes, you removed the templates/revert-button/index.tsx path from the side-effects array configuration in package.json. That file imports the @wordpress/site-editor package (and is the only file in all of Woo to do so). Likely not a coincidence and also indicates that the revert button functionality might not be working as expected in production builds (I don't see it in your testing notes).

Interesting! That looks like the cause of the problem indeed, but I wonder what is the correct approach to follow, because based on the release notes for WordPress 6.3, importing from `'@wordpress/edit-site'` is the recommended approach for adding additional information in the site editor’s Template sidebar, and this is what we are doing on `/assets/js/templates/revert-button/index.tsx`: https://make.wordpress.org/core/2023/07/18/miscellaneous-editor-changes-in-wordpress-6-3/

gigitux commented:
> > I notice that in your changes, you removed the templates/revert-button/index.tsx path from the side-effects array configuration in package.json. That file imports the @wordpress/site-editor package (and is the only file in all of Woo to do so). Likely not a coincidence and also indicates that the revert button functionality might not be working as expected in production builds (I don't see it in your testing notes).
> 
> Interesting! That looks like the cause of the problem indeed, but I wonder what is the correct approach to follow, because based on the release notes for WordPress 6.3, importing from `'@wordpress/edit-site'` is the recommended approach for adding additional information in the site editor’s Template sidebar, and this is what we are doing on `/assets/js/templates/revert-button/index.tsx`: https://make.wordpress.org/core/2023/07/18/miscellaneous-editor-changes-in-wordpress-6-3/

It is the only way to import that function :/ 

We should try to import the package with a dynamic import. I tried, but I wasn't able to do it 🤔

Aljullu committed:
Refactor logic that sets default attributes to Products block depending on the template

Aljullu committed:
Refactor code that was checking whether the Mini-Cart block was being edited inside a template

nefeline commented:
> We should try to import the package with a dynamic import. I tried, but I wasn't able to do it 🤔

Yeah, it's tricky. I was trying to figure out alternative approaches, but nothing came to mind up until now. Should we work on the revert button on a separate PR to unblock the patch release?

gigitux commented:
> Yeah, it's tricky. I was trying to figure out alternative approaches, but nothing came to mind up until now. Should we work on the revert button on a separate PR to unblock the patch release?

We have times to fix the issue (I would say at least one week), is there a particular reason you want to do a patch release now?

(I would prefer reduce the number of patch release)

nefeline commented:
> We should try to import the package with a dynamic import. I tried, but I wasn't able to do it 🤔

@gigitux ~do you mind sharing more details about the approach you attempted to implement for a dynamic import?~

Edit: I just noticed that when you attempted to solve this problem back in May, [Riad mentioned over here](https://github.com/WordPress/gutenberg/pull/50257#issuecomment-1549282882) that we: 

> might be "bundling" the edit-site dependency which duplicates some JS packages

[Did you make any progress](https://github.com/WordPress/gutenberg/pull/50257#issuecomment-1549298733) on that front?

nefeline commented:
Update regarding the `revert-button`:

I made some attempts to solve the problem, including:

* Conditionally loading `'@wordpress/edit-site'` via PHP;
* Dynamic import to load modules on-demand, at runtime, instead of statically at the beginning.

But none of those worked. We are following [the recommendations from Gutenberg](https://github.com/WordPress/gutenberg/pull/50257) for the usage of the `PluginTemplateSettingPanel` component so I reached out to Nik on https://github.com/WordPress/gutenberg/pull/50257#issuecomment-1648284471 to ask a few questions about not only this conflict with the navigation block, but also regarding the `core/interface is already registered` error.

nefeline committed:
Merge branch 'trunk' into 10259-woocommerce-causing-malfunction-of-the-navigation-block-on-wordpress-63

nefeline committed:
Ensure the 'wp-edit-site' script is enqueued in the site editor context exclusively.

nefeline commented:
Update: the problem is now fixed [b44fecd](https://github.com/woocommerce/woocommerce-blocks/pull/10283/commits/b44fecd5db7c7e84e1ccf5b5b81386fcc21d0a7d) ! 🎉 

## Investigation Details

After Darren helped me out with [narrowing down the cause of this problem to the revert button](https://github.com/woocommerce/woocommerce-blocks/pull/10283#discussion_r1271443132) and the import of the `@wordpress/site-editor` package, I started to investigate ways to [conditionally load this package via PHP](https://github.com/woocommerce/woocommerce-blocks/pull/10283#discussion_r1272343843).

In parallel, I've also reached out to the GB team, and [Jarda shared an excellent explanation](https://github.com/WordPress/gutenberg/pull/50257#issuecomment-1651200732) with more details about the reasons why we were facing this problem. He also suggested the following approach for solving it:

> Woo Blocks should build templates/revert-button as a separate script. It's an independent standalone plugin. And then enqueue it only in the Site Editor, nowhere else. That should happen in PHP code, somewhere in AssetsController.php.

Which served as a confirmation that I was on the right path when choosing to conditionally load the site-editor script via PHP.

## The solution

After further investigation, I found out that, as it turns out, WebPack has a [configuration option called `externals`](https://webpack.js.org/configuration/externals/), which excludes certain modules from the final bundle, and those modules can then be expected to be available in the global scope at runtime. 

With `externals`, we can exclude specific modules from the final bundle, and those modules will then be expected to be available in the global scope at runtime. This is very handy to conditionally load something based on the presence of a global variable.

With that said, the solution was implemented in three parts:

### 1. Updated the `webpack.config.js` file to include:

```js
module.exports = {
  // ... other webpack config ...

  externals: {
    '@wordpress/edit-site': 'wp.editSite'
  }
};
```
Here, we're telling Webpack not to bundle `@wordpress/edit-site` but to expect it to be available globally as `wp.editSite`.

### 2. Conditionally Loaded `'wp-edit-site'` via PHP:

On PHP, it was necessary to specify the script to run on the 'site-editor' screen specifically as `'enqueue_block_editor_assets'` is used to enqueue assets for more areas, such as when editing posts, pages, etc.

```php
add_action( 'enqueue_block_editor_assets', 'enqueue_site_editor_dependencies' );
	
function enqueue_site_editor_dependencies() {
	$current_screen = get_current_screen();

	if ( $current_screen instanceof \WP_Screen && 'site-editor' === $current_screen->base ) {
		wp_enqueue_script( 'wp-edit-site' );
	}
}
```

### 3. Imported `@wordpress/edit-site` as usual:

Even though the module is marked as external, it is still possible to import it into our components. With the solution implemented here, Webpack will simply replace these imports with references to the global variable (which is `wp.editSite` in this case):

```js
import editSite from '@wordpress/edit-site';
```

nerrad commented on bin/webpack-entries.js:
```diff
@@ -178,4 +178,8 @@ const getEntryConfig = ( type = 'main', exclude = [] ) => {
 module.exports = {
 	getEntryConfig,
+
+	externals: {
+		'@wordpress/edit-site': 'wp.editSite',
+	},
```
`@wordpress/edit-site` is [already setup as an external](https://github.com/WordPress/gutenberg/blob/fa4f5a5e4e5a1c1e1309f59ff9a99526f92a0a6a/packages/dependency-extraction-webpack-plugin/lib/util.js#L45-L50) via `@wordpress/dependency-extraction-webpack-plugin`. This is unnecessary.

nerrad reviewed (changes requested):
What Jarda was getting at in his comment on the Gutenberg ticket is that we should be building a different entrypoint for the script that requires `wp.edit-site` so that script gets conditionally loaded in the site editor context. Right now `revert-button`, is a direct dependent of the `wc-blocks-shared-context` entry point and it shouldn't be.

A few concerns about the latest changes:

- The extra configuration for `@wordpress/edit-site` as an external is not needed. That's handled already in our configuration.
- The changes to enqueueing `wc-blocks-shared-context` with no dependencies is fragile (it should utilize any `assets.php` file if it exists. The current implementation is fragile if in the future a dependency _is_ added, it could easily be missed to have that dependency registered here.
- Explicitly enqueuing `wp-edit-site` should be unnecessary. If we break out a new entry point script for `revert_button`, that script registration only on the site editor context should automatically take care of the correct dependencies assuming we're using the context of its relevant `assets.php` for getting them.

nefeline commented:
My apologies @nerrad :  I should have moved this PR to draft/in progress as I'm still figuring things out: I was experimenting with webpack externals and a few other ideas: will ping you as soon as this is ready.

nefeline commented:
Closing this out in favor of https://github.com/woocommerce/woocommerce-blocks/pull/10388
